wordpress-mobile/WordPress-FluxC-Android
# Woo/fetch product categories

Fixes #1587 by adding support to fetch product categories for a given site.

#### Changes
- Adds a new endpoint `wc/v3/products/categories/` to fetch product categories for a site.
- Added new `WCProductCategoryModel` model class and a migration script to create table. 
- Added new action and action classes to `WCProductStore` and `ProductRestClient` to fetch product categories for a site.
- Added unit tests.
- Added a new button to the example app to fetch product categories.

#### Notes
- ~**This PR is in draft till this [Shipping Labels PR](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/1585) can be reviewed and merged.**~
- This work has largely pulled changes from this [PR](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/1569) but rather than submitting it as a single PR, I chose to split the PR into smaller chunks and clean up the examples and unit tests in order for better review.
- The option to update a product category and add a product category will be implemented in subsequent PRs.

#### Screenshots
<img src="https://user-images.githubusercontent.com/22608780/82629753-42a7f480-9c0e-11ea-9256-206d78ed48f4.png" width="300"/>. <img src="https://user-images.githubusercontent.com/22608780/82629637-f8bf0e80-9c0d-11ea-9168-9db188b979e5.gif" width="300"/>

#### To test
- Since this PR adds a migration script, it would be nice to test updating the app.
- Run `MockedStack_WCProductsTest` and `ReleaseStack_WCProductTest` and `ProductSqlUtilsTest`.
- In the example app, click on `Woo` -> `Products` - > `Select Site` - > `Fetch product categories`.
  - Verify that the categories are fetched correctly.
  - Click on `Lore more categories` and verify that the categories are fetched correctly.
  - If there is no more categories data found in the API response, the `Load more categories` button should NOT be displayed.
